smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/SymbolVisitor.java

```diff
@@ -378,18 +396,63 @@ private String flattenShapeName(ToShapeId id) {
                 .definitionFile(toFilename(namespace));
     }
 
-    private String formatModuleName(ShapeType shapeType, String name) {
-        // All shapes except for the service and operations are stored in models.
-        if (shapeType == ShapeType.SERVICE) {
-            return "./" + name;
-        } else if (shapeType == ShapeType.OPERATION) {
-            return "./commands/" + name;
-        } else {
-            return "./models/index";
-        }
-    }
-
     private String toFilename(String namespace) {
         return namespace + ".ts";
     }
+
+    /**
+     * Utility class to locate which path should the symbol be generated into.
+     * It will break the models into multiple files to prevent it being too big.
+     */
+    static final class ModuleNameDelegator {
+        static final int DEFAULT_CHUNK_SIZE = 300;
+        static final String SHAPE_NAMESPACE_PREFIX = "./models/";
+
+        private final Map<Shape, String> visitedModels = new HashMap<>();
+        private int bucketCount = 0;
+        private int currentBucketSize = 0;
+        private final int chunkSize;
+
+        ModuleNameDelegator(int shapeChunkSize) {
+            chunkSize = shapeChunkSize;
+        }
+
+        ModuleNameDelegator() {
+            chunkSize = DEFAULT_CHUNK_SIZE;
+        }
+
+        public String formatModuleName(Shape shape, String name) {
+            // All shapes except for the service and operations are stored in models.
+            if (shape.getType() == ShapeType.SERVICE) {
+                return "./" + name;
+            } else if (shape.getType() == ShapeType.OPERATION) {
+                return "./commands/" + name;
+            } else {
+                if (visitedModels.containsKey(shape)) {
+                    return visitedModels.get(shape);
+                }
+                if (currentBucketSize == chunkSize) {
+                    bucketCount++;
+                    currentBucketSize = 0;
+                }
+                currentBucketSize++;
+                String path = SHAPE_NAMESPACE_PREFIX + "models_" + bucketCount;
+                visitedModels.put(shape, path);
+                return path;
+            }
+        }
+
+        static void writeModelIndex(Model model, SymbolProvider symbolProvider, FileManifest fileManifest) {
+            TypeScriptWriter writer = new TypeScriptWriter("");
+            List<String> namespaces = model.shapes().map(shape -> symbolProvider.toSymbol(shape).getNamespace())
+                    .filter(namespace -> namespace.startsWith(SHAPE_NAMESPACE_PREFIX))
+                    .collect(Collectors.toSet())
+                    .stream()
+                    .collect(Collectors.toList());
+            namespaces.sort(Comparator.naturalOrder());
+            namespaces.forEach(namespace -> writer.write(
```

Comment:
This can all be simplified to only use streams:

```
model.shapes()
        .map(shape -> symbolProvider.toSymbol(shape).getNamespace())
        .filter(namespace -> namespace.startsWith(SHAPE_NAMESPACE_PREFIX))
        .distinct()
        .sort(Comparator.naturalOrder())
        .forEach(X -> {})
```

Reply:
Looks like I cannot chain the `forEach` after `.sort()`. So I will do:

```java
            List<String> namespaces = model.shapes()
                    .map(shape -> symbolProvider.toSymbol(shape).getNamespace())
                    .filter(namespace -> namespace.startsWith(SHAPE_NAMESPACE_PREFIX))
                    .distinct()
                    .collect(Collectors.toList());
            namespaces.sort(Comparator.naturalOrder());
            namespaces.forEach(namespace -> writer.write(
                        "export * from $S;", namespace.replaceFirst(SHAPE_NAMESPACE_PREFIX, "./")));
```

Reply:
Used `sorted()` in streams instead